### PR TITLE
PIM-4446: manipulate common objects

### PR DESCRIPTION
**Pull request by nidup**
don't forget to merge https://github.com/akeneo/pim-community-dev/tree/doc-updater-api2 to revert paginator changes that we don't document for now

### CODE REVIEW

File: cookbook/catalog/common/create.rst [hunk 1]
```diff
@@ -0,0 +1,58 @@
+How to Create Non-Product Objects
+=================================
+
+To create common objects, we rely on different methods depending on the complexity of the object, there is few examples on how to use these services.
```
Comment (jjanvier) [inline cookbook/catalog/common/create.rst]: there are

File: cookbook/catalog/common/query.rst [hunk 2]
```diff
@@ -0,0 +1,67 @@
+How to Query Non-Product Objects
+================================
+
+Prerequisites
+-------------
+
+The Akeneo PIM project introduces services to help you manage your objects.
+
+Product is the main object of the application and had a large set of services to build and execute queries.
```
Comment (jjanvier) [inline cookbook/catalog/common/query.rst]: has a large

File: cookbook/catalog/common/create.rst [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+.. note::
+
+    To ease the override of any object in your projects, we avoid any direct use of 'new MyObject()' in our code. We rely on services which implement a "creational pattern" and where we inject a config parameter, for instance, 'pim_catalog.entity.family.class'.
```
Comment (jmleroux) [inline cookbook/catalog/common/create.rst]: not 'and' in ' and where we inject...'

File: cookbook/catalog/common/create.rst [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    // fetch the factory from the container
+    $factory = $this->getContainer()->get('pim_catalog.factory.attribute');
+
+    // create an attribute with the attribute type
+    $attribute = $factory->createAttribute('pim_catalog_text');
+    $attribute->setCode('mynewattribute');
+
+.. note::
+
+    ``Pim\Bundle\CatalogBundle\Resources\config\attribute_types.yml`` contains the list of default attribute types, the alias of a service tagged with 'pim_catalog.attribute_type' is an attribute type useable here, for instance, 'pim_catalog_identifier', 'pim_catalog_text', 'pim_catalog_textarea', etc
```
Comment (jmleroux) [inline cookbook/catalog/common/create.rst]: and we should have constants for the attributes types imho ;)

File: cookbook/catalog/common/query.rst [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+For all other simpler objects (as Attribute, Family, Category, Channel, Locale etc), we use quite common Doctrine repositories.
+
+Instantiate an ObjectRepository
+-------------------------------
+
+Doctrine is very used in the application for persisting and reading information to and from a database.
+
+For each object, a repository which implements ``Doctrine\Common\Persistence\ObjectRepository`` is provided.
```
Comment (jmleroux) [inline cookbook/catalog/common/query.rst]: `extend` instead of ìmplement` ?

File: cookbook/catalog/common/query.rst [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+These classes embed the logic to build and execute queries and are defined as services that you can fetch from the container.
+
+.. code-block:: php
+
+    $attributeRepository = $this->container->get('pim_catalog.repository.attribute');
+    $categoryRepository = $this->container->get('pim_catalog.repository.category');
+    $localeRepository = $this->container->get('pim_catalog.repository.locale');
```
Comment (jmleroux) [inline cookbook/catalog/common/query.rst]: align assignments (Akeneo coding best pratices)

File: cookbook/catalog/common/query.rst [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+Each dedicated repository also implements its own business interface, for instance, ``Pim\Bundle\CatalogBundle\Repository\AttributeRepositoryInterface``.
+
+This business interface provides several dedicated methods and we strongly advise to rely on this interface in your developments.
+
+.. code-block:: php
+
+    $attributeCode = $attributeRepository->getIdentifierCode();
+    $categories = $categoryRepository->getAllChildrenIds($myCategory);
+    $locales = $localeRepository->getActivatedLocales();
```
Comment (jmleroux) [inline cookbook/catalog/common/query.rst]: align

File: cookbook/catalog/common/update.rst [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+Here we can apply many property updates on any fields of an attribute.
+
+.. code-block:: php
+
+    $updater = $this->getContainer()->get('pim_catalog.updater.attribute');
+    $fieldUpdates = [
+        'labels' => ['en_US' => 'Test1', 'fr_FR' => 'Test2'],
+        'group' => 'marketing',
+        'attributeType' => 'pim_catalog_text'
```
Comment (jmleroux) [inline cookbook/catalog/common/update.rst]: align

File: cookbook/catalog/common/create.rst (hunk 3 above)
Comment (phaseinducer) [inline cookbook/catalog/common/create.rst]: just "configurable services" ?

File: cookbook/catalog/common/update.rst [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+The Akeneo PIM project introduces services to help you to manage your business objects.
+
+The update of an object can be done through the use of classic setters directly on the model.
+
+The version 1.4 introduces a new internal API to ease these updates, it especially helps when you need to change relation between objects because it relies on unique code of related objects.
+
+These updaters implements a ``Akeneo\Component\StorageUtils\Updater\ObjectUpdaterInterface`` to apply a set of property updates on an object.
+
+We strongly advise to use this internal API for your developments and there is few examples on how to use these dedicated updaters.
```
Comment (jjanvier) [inline cookbook/catalog/common/update.rst]: and here are a few examples

File: cookbook/catalog/common/validate.rst [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+Use the Validator
+-----------------
+
+The validator is a service, you can fetch it from the container
+
+.. code-block:: php
+
+    $validator = $this->getContainer()->get('validator');
+
+When you can validate an attribute, a family or another object, all business validation are applied.
```
Comment (jjanvier) [inline cookbook/catalog/common/validate.rst]: validationS

File: cookbook/catalog/common/query.rst [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+The Akeneo PIM project introduces services to help you manage your objects.
+
+Product is the main object of the application and had a large set of services to build and execute queries.
+
+For all other simpler objects (as Attribute, Family, Category, Channel, Locale etc), we use quite common Doctrine repositories.
+
+Instantiate an ObjectRepository
+-------------------------------
+
+Doctrine is very used in the application for persisting and reading information to and from a database.
```
Comment (phaseinducer) [inline cookbook/catalog/common/query.rst]: "widely used" sounds better

File: cookbook/catalog/common/query.rst [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+    // Finds objects by a set of criteria.
+    public function findBy(array $criteria, array $orderBy = null, $limit = null, $offset = null);
+
+    // Finds a single object by a set of criteria.
+    public function findOneBy(array $criteria);
+
+.. note::
+
+    Don't hesitate to take a look on Doctrine documentation for further information
```
Comment (phaseinducer) [inline cookbook/catalog/common/query.rst]: final period .

File: cookbook/catalog/common/save.rst [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+===============================
+
+To save one or many objects, we provide a dedicated service which provides methods 'save' and 'saveAll' through the implementation of ``Akeneo\Component\StorageUtils\Saver\SaverInterface`` and ``Akeneo\Component\StorageUtils\Saver\BulkSaverInterface``.
+
+Use the Saver to Save a Single Object
+-------------------------------------
+
+You can save one or many objects of a kind with a dedicated service, the saver checks that the used object is supported (for instance, you can't use the attribute saver to save a family).
+
+We define these different services to ease the future changes and allow you to override only one of them to add a specific business logic.
```
Comment (phaseinducer) [inline cookbook/catalog/common/save.rst]: "a" is useless here

File: cookbook/catalog/common/save.rst [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+    $categorySaver = $this->getContainer()->get('pim_catalog.saver.category');
+    $categorySaver->saveAll([$categoryOne, $categoryTwo]);
+
+Use the Saver with Options
+--------------------------
+
+You can use the following extra boolean options as second parameter when you save objects.
+
+If the 'flush' option is passed with 'true', the object will be saved in database.
```
Comment (phaseinducer) [inline cookbook/catalog/common/save.rst]: Maybe it's a good place to say what is the default value.

File: cookbook/catalog/common/query.rst (hunk 5 above)
Comment (nidup) [inline cookbook/catalog/common/query.rst]: ObjectRepository is an interface ... Doctrine stuff

File: cookbook/catalog/common/query.rst (hunk 6 above)
Comment (nidup) [inline cookbook/catalog/common/query.rst]: ok ;)

File: cookbook/catalog/common/create.rst (hunk 4 above)
Comment (nidup) [inline cookbook/catalog/common/create.rst]: yes we should :)

File: cookbook/catalog/common/query.rst [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+    $attributeRepository = $this->container->get('pim_catalog.repository.attribute');
+    $categoryRepository = $this->container->get('pim_catalog.repository.category');
+    $localeRepository = $this->container->get('pim_catalog.repository.locale');
+    // ...
+
+Use an ObjectRepository
+-----------------------
+
+Each repository dedicated to a commnon object provides methods from the interface ``Doctrine\Common\Persistence\ObjectRepository``.
```
Comment (grena) [inline cookbook/catalog/common/query.rst]: common* nom nom nom

File: cookbook/catalog/common/save.rst (hunk 14 above)
Comment (nidup) [inline cookbook/catalog/common/save.rst]: yes it is

File: cookbook/catalog/common/query.rst (hunk 12 above)
Comment (grena) [inline cookbook/catalog/common/query.rst]: take a look at*

File: cookbook/catalog/common/query.rst [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+.. code-block:: php
+
+    $attributeCode = $attributeRepository->getIdentifierCode();
+    $categories = $categoryRepository->getAllChildrenIds($myCategory);
+    $locales = $localeRepository->getActivatedLocales();
+    // ...
+
+.. tip::
+
+    In your own developments, you should always put the code which build and execute queries in your own repositories.
```
Comment (grena) [inline cookbook/catalog/common/query.rst]: which build**s** and execute**s**

File: cookbook/catalog/common/remove.rst [hunk 17]
```diff
@@ -0,0 +1,38 @@
+How to Remove Non-Product Objects
+=================================
+
+To remove one or many objects, we provide a dedicated service which provides methods 'remove' and 'removeAll' through the implementation of ``Akeneo\Component\StorageUtils\Remover\RemoverInterface`` and ``Akeneo\Component\StorageUtils\Remover\BulkRemoverInterface``.
+
+Use the Remover to Remove a Single Object
```
Comment (grena) [inline cookbook/catalog/common/remove.rst]: the Remover to **r**emove a

File: cookbook/catalog/common/remove.rst [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+-----------------------------------------
+
+You can remove one or many objects of a kind with a dedicated service, the remover checks that the used object is supported (for instance, you can't use the attribute remover to remove a family).
+
+We define these different services to ease the future changes and allow you to override only one of them to add a specific business logic.
+
+Some services already use specific classes but most of these services use the class ``Akeneo\Bundle\StorageUtilsBundle\Doctrine\Common\Remover\BaseRemover``.
```
Comment (grena) [inline cookbook/catalog/common/remove.rst]: By "use", you mean "extend"? If so, I thing we should precise :)

File: cookbook/catalog/common/remove.rst [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    $attributeRemover = $this->getContainer()->get('pim_catalog.remover.attribute');
+    $attributeRemover->remove($attribute);
+
+    $familyRemover = $this->getContainer()->get('pim_catalog.remover.family');
+    $familyRemover->remove($family);
+
+    $categoryRemover = $this->getContainer()->get('pim_catalog.remover.category');
+    $categoryRemover->remove($category);
+
+Use the Remover to Remove many Objects
```
Comment (grena) [inline cookbook/catalog/common/remove.rst]: Use the Remover to **r**emove many Objects

File: cookbook/catalog/common/update.rst [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+
+.. warning::
+
+   The updater does not validate and save the products in the database, these operations are done by the Validator and the Saver (detailed in specific chapters).
```
Comment (phaseinducer) [inline cookbook/catalog/common/update.rst]: :+1: 
This sentence will save lots of pain for the community managers. :P

File: cookbook/catalog/common/save.rst [hunk 21]
```diff
@@ -0,0 +1,53 @@
+How to Save Non-Product Objects
+===============================
+
+To save one or many objects, we provide a dedicated service which provides methods 'save' and 'saveAll' through the implementation of ``Akeneo\Component\StorageUtils\Saver\SaverInterface`` and ``Akeneo\Component\StorageUtils\Saver\BulkSaverInterface``.
+
+Use the Saver to Save a Single Object
```
Comment (grena) [inline cookbook/catalog/common/save.rst]: Use the Saver to **s**ave a **s**ingle Object

File: cookbook/catalog/common/save.rst [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+-------------------------------------
+
+You can save one or many objects of a kind with a dedicated service, the saver checks that the used object is supported (for instance, you can't use the attribute saver to save a family).
+
+We define these different services to ease the future changes and allow you to override only one of them to add a specific business logic.
+
+Some services already use specific classes but most of these services use the class ``Akeneo\Bundle\StorageUtilsBundle\Doctrine\Common\Saver\BaseSaver``.
```
Comment (grena) [inline cookbook/catalog/common/save.rst]: Same remark for the word "use"

File: cookbook/catalog/common/update.rst [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+            'fr_FR' => 'Moniteurs',
+            'en_US' => 'PC Monitors',
+        ]
+    ];
+    $this->updater->update($family, $fieldUpdates);
+
+Use the CategoryUpdater to Update a Category
+--------------------------------------------
+
+Here we can apply many property updates on any fields of a category.
```
Comment (phaseinducer) [inline cookbook/catalog/common/update.rst]: many => several

File: cookbook/catalog/common/save.rst (hunk 14 above)
Comment (grena) [inline cookbook/catalog/common/save.rst]: What if the behaviour if set to `false`?

File: cookbook/catalog/product/save.rst [hunk 24]
```diff
@@ -13,7 +13,7 @@ The product saver is a service, you can fetch it from the container.
 Save the Products
 -----------------
 
-It implements SaverInterface and BulkSaverInterface so you can save one or many products.
+It implements ``Akeneo\Component\StorageUtils\Saver\SaverInterface`` and ``Akeneo\Component\StorageUtils\Saver\BulkSaverInterface`` so you can save one or many products.
```
Comment (phaseinducer) [inline cookbook/catalog/product/save.rst]: many => several

File: cookbook/catalog_structure/index.rst [hunk 25]
```diff
@@ -1,9 +1,12 @@
 How to Customize the Catalog Structure
 ======================================
 
+This chapter gives details on how customizing the catalog structure to add your own models or update existing one.
```
Comment (phaseinducer) [inline cookbook/catalog_structure/index.rst]: oneS

File: cookbook/catalog/common/update.rst [hunk 26]
```diff
@@ -0,0 +1,80 @@
+How to Update Non-Product Objects
+=================================
+
+Prerequisites
+-------------
+
+The Akeneo PIM project introduces services to help you to manage your business objects.
```
Comment (grena) [inline cookbook/catalog/common/update.rst]: to help you manage

File: cookbook/catalog/common/update.rst [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+
+    $updater = $this->getContainer()->get('pim_catalog.updater.category');
+    $fieldUpdates = [
+        'code'         => 'mycode',
+        'parent'       => 'master',
+        'labels'       => [
+            'fr_FR' => 'Ma superbe catégorie',
+        ],
+    ];
+    $this->updater->update(category, $fieldUpdates);
```
Comment (grena) [inline cookbook/catalog/common/update.rst]: **$**category

File: cookbook/catalog/common/update.rst [hunk 28]
```diff
@@ -0,0 +1,80 @@
...
+
+.. tip::
+
+    The same approach is used for other business objects, to know all the updatable fields you can take a look on our Specs, for instance, ``spec\Pim\Component\Catalog\Updater\\CategoryUpdaterSpec``
```
Comment (grena) [inline cookbook/catalog/common/update.rst]: there is an extra backslash in the FQCN

File: cookbook/catalog/common/validate.rst (hunk 10 above)
Comment (grena) [inline cookbook/catalog/common/validate.rst]: I don't get the "When you **_can_** validate (...)", why not "When you validate an attribute (...)"?

File: cookbook/catalog/common/validate.rst [hunk 29]
```diff
@@ -0,0 +1,27 @@
...
+
+.. code-block:: php
+
+    $violations = $validator->validate($family);
+    $violations = $validator->validate($attribute);
+    // ...
+
+It returns here a list of violation errors, a 'Symfony\Component\Validator\ConstraintViolationList'.
```
Comment (grena) [inline cookbook/catalog/common/validate.rst]: I think you can use a double back quote to "codify" the name : 
```
``SymfonyComponentValidatorConstraintViolationList``
```

File: cookbook/catalog/common/validate.rst [hunk 30]
```diff
@@ -0,0 +1,27 @@
...
+
+.. warning::
+
+    The validation of a product is a bit different, you should use its own validator, the service alias is  `pim_catalog.validator.product`.
```
Comment (grena) [inline cookbook/catalog/common/validate.rst]: same here, to double-back-quotify \o/

File: cookbook/catalog/common/create.rst [hunk 31]
```diff
@@ -0,0 +1,58 @@
...
+--------------------------------------------
+
+The ``Pim\Bundle\CatalogBundle\Factory\FamilyFactory`` is available as a service, you can fetch it from the container.
+
+.. code-block:: php
+
+    // fetch the factory from the container
+    $factory = $this->getContainer()->get('pim_catalog.factory.family');
+
+    // create an family, internally, the attribute requirement for the identifier attribute is directly added
```
Comment (willy-ahva) [inline cookbook/catalog/common/create.rst]: create A family

File: cookbook/catalog/common/create.rst [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+------------------------------------------------
+
+The ``Pim\Bundle\CatalogBundle\Factory\CategoryFactory`` is available as a service, you can fetch it from the container.
+
+.. code-block:: php
+
+    // fetch the factory from the container
+    $factory = $this->getContainer()->get('pim_catalog.factory.category');
+
+    // create an category
```
Comment (willy-ahva) [inline cookbook/catalog/common/create.rst]: create A category

File: cookbook/catalog/common/remove.rst [hunk 33]
```diff
@@ -0,0 +1,38 @@
...
+=================================
+
+To remove one or many objects, we provide a dedicated service which provides methods 'remove' and 'removeAll' through the implementation of ``Akeneo\Component\StorageUtils\Remover\RemoverInterface`` and ``Akeneo\Component\StorageUtils\Remover\BulkRemoverInterface``.
+
+Use the Remover to Remove a Single Object
+-----------------------------------------
+
+You can remove one or many objects of a kind with a dedicated service, the remover checks that the used object is supported (for instance, you can't use the attribute remover to remove a family).
+
+We define these different services to ease the future changes and allow you to override only one of them to add a specific business logic.
```
Comment (willy-ahva) [inline cookbook/catalog/common/remove.rst]: I don't really understand what it means => `allow you to override only one of them`

File: cookbook/catalog/common/save.rst [hunk 34]
```diff
@@ -0,0 +1,53 @@
...
+===============================
+
+To save one or many objects, we provide a dedicated service which provides methods 'save' and 'saveAll' through the implementation of ``Akeneo\Component\StorageUtils\Saver\SaverInterface`` and ``Akeneo\Component\StorageUtils\Saver\BulkSaverInterface``.
+
+Use the Saver to Save a Single Object
+-------------------------------------
+
+You can save one or many objects of a kind with a dedicated service, the saver checks that the used object is supported (for instance, you can't use the attribute saver to save a family).
+
+We define these different services to ease the future changes and allow you to override only one of them to add specific business logic.
```
Comment (willy-ahva) [inline cookbook/catalog/common/save.rst]: Same remark for `only one of them`

File: cookbook/catalog/common/remove.rst (hunk 17 above)
Comment (nidup) [inline cookbook/catalog/common/remove.rst]: http://grammar.yourdictionary.com/capitalization/rules-for-capitalization-in-titles.html ?

File: cookbook/catalog/common/remove.rst (hunk 33 above)
Comment (grena) [inline cookbook/catalog/common/remove.rst]: As there are one identified service per managed item, you can override only one service (for product, or category..) etc

Comment (nidup) [inline cookbook/catalog/common/remove.rst]: means only the family remover but not the attribute remover

File: cookbook/catalog/common/remove.rst (hunk 18 above)
Comment (nidup) [inline cookbook/catalog/common/remove.rst]: really use

File: cookbook/catalog/common/save.rst (hunk 34 above)
Comment (nidup) [inline cookbook/catalog/common/save.rst]: i add a comment, thx

File: cookbook/catalog/common/update.rst (hunk 27 above)
Comment (nidup) [inline cookbook/catalog/common/update.rst]: :eyes: of the :tiger:

File: cookbook/catalog/common/update.rst (hunk 28 above)
Comment (nidup) [inline cookbook/catalog/common/update.rst]: :eyes: of the :tiger:

File: cookbook/catalog/common/validate.rst (hunk 29 above)
Comment (nidup) [inline cookbook/catalog/common/validate.rst]: good idea

**Pull request merged**